### DISCUSSION TITLE
Item not available (Still)

### DISCUSSION THREAD

**Discussion by GameModes** [gijzelaerr/python-snap7#362] [Q&A] [answered]
Good day,
I want to connect Python with my S7 PLC machine, but I have received some (security) errors when I enter some of the default codes.
This is the error:

`
b'CPU: Item not available'
`

This happens for at least: read_area, get_protection & get_plc_datetime

I found out this Issue is a security issue with this discussion and followed it: https://github.com/gijzelaerr/python-snap7/discussions/323 and turned on "Permit acces with PUT/GET communication from remote partner" with Tia portal (which I'm not an expert on) and downloaded it to my device. Still this didn't help. Though I'm not sure if the "Permit Access" setting is truly on the PLC. I have compiled and downloaded it on the device and more, I dont know.
Any ideas what to do next? Or is this the final answer? https://github.com/gijzelaerr/python-snap7/discussions/323#discussioncomment-2356784

with kind regards,
Ruben

**Comment by swamper123** [comment]
Hey Ho :)

`b'CPU: Item not available'` can be caused either:

- The DB was wrongly configured

or

-  You try to write/read unsupported types

What have to be configured and what is supported is shown [here](http://snap7.sourceforge.net/snap7_client.html#target_compatibility). These configs have to be done and loaded on the PLC.

It seems that snap7 was installed and loaded correctly, since you received an error message from snap7 and it isn't a Pythonerror which shows up (that would be our fault then).

 `read_area` should work on all devices.

**Comment by GameModes** [comment, marked as answer]
Hey,

"The DB was wrongly configured" is probably the fault.
because if "read_area should work on all devices." then my command:
`
client.read_area(snap7.types.Areas.DB, 1, 10, 4)

`
is because of the DB. Is there a command for read_area that doesnt involve db that I can use?
And the "DB" is that the database implemented in the PLC siemens machine or something else?

Because my goal is to read a value from the PLC with their PLC address like: "M0" or "DB24.20" and get the current value of it.
But I cant figure out the correct command to receive such data.

**Reply by swamper123** [reply]
tldr: If no DB exists on the PLC, you will not be able to read from it.


There are different types of areas where you can read from: Counter, Timer, DB ("so called Daten Baustein" in german)... I guess PE are Inputs ("Prozess Eingänge"), PA Outputs ("Prozess Ausgänge") and MK are "Merker".

Mostly for data, a so called "DB" Block exists, where values are stored, like in a SQL Database.
To be honest, I have never tested MK nor PE nor PA with a real siemens. Mostly I wanted a defined processvalue with a nice datatype I can work with.

This "DB" has to exist on the PLC of course:

From your example you say:
`client.read_area(snap7.types.Areas.DB, 1, 10, 4)`

"Dear Siemens PLC, from DB1, I want to read, starting from Address 10.0, 4 Bytes." (see in the method description as well ;) )
Then you get the raw data. There are plenty of util methods to convert them into sth. more usefull (and more are incoming with the newest PR, I hope :wink:  ).
If you want to make a programm, which uses the "PLC Name", you have to parse them on your own. But remember, that if you use blank `read_area(...)` you have to know how many bytes you have to read. 

But how to build such a thing is another story, which is out of the scope of this repo.

**Reply by GameModes** [reply]
So if I want to read data with a certain address like DB42.23 the command is?:
`
client.read_area(DB42.23, *other parameters)
`
Or do you mean that, getting raw data from a plc address is not possible with the repo?

**Reply by swamper123** [reply]
If you want to read DB42.23 you would need sth. like this:

`client.read_area(snap7.types.Areas.DB, 42, 23, AMOUNT_OF_BYTES_YOU_WANT_TO_READ)`

`AMOUNT_OF_BYTES_YOU_WANT_TO_READ` is the size of your datatype in the DB at this address (which is the usual case).

**Reply by GameModes** [reply]
I'm a bit confused with the output that I get. So I do the following (I still dont really understand the amount_of_bytes part but I just use 2):

`
client.read_area(snap7.types.Areas.DB, 23, 88, 2)
`

And the value should be "2"
but instead I get this weird bytearray:

`
bytearray(b'\x00\x00')
`

How can I convert it to a normal number?

**Reply by swamper123** [reply]
Yes, if you use `read_area()` you always get a bytearray back.
If you take a look here: https://github.com/gijzelaerr/python-snap7/blob/master/snap7/util.py
You can find many helper functions to make it more "normal stuff" like.

But it looks like, that these two bytes starting from DB23.88 contains zeros. You may check your DB again.
What kind of datatype should it be? INT? DINT? ...

**Reply by GameModes** [reply]
INT

**Reply by GameModes** [reply]
Will this function convert the bytearray to a number?:
![image](https://user-images.githubusercontent.com/60732314/160076586-9992d497-8813-4d91-bccd-4d9d27804bfb.png)

**Reply by swamper123** [reply]
Get byte takes a bytearray and an index and returns an int (seems to be ... strange tbh.)


I would try this one:

https://github.com/gijzelaerr/python-snap7/blob/b278707d750522f0f465dd4ac101703a186f284d/snap7/util.py#L282

**Reply by swamper123** [reply]
Description of this method, you  can find in the comments below (of the method).
